c11threads_win32.c

```diff
@@ -0,0 +1,801 @@
+/*
+Win32 implementation for c11threads.
+
+Authors:
+John Tsiombikas <nuclear@member.fsf.org>
+Oliver Old <oliver.old@outlook.com>
+
+I place this piece of code in the public domain. Feel free to use as you see
+fit. I'd appreciate it if you keep my name at the top of the code somewhere, but
+whatever.
+
+Main project site: https://github.com/jtsiomb/c11threads
+*/
+
+#if defined(_WIN32) && !defined(C11THREADS_PTHREAD_WIN32)
+
+/* Map debug malloc and free functions for debug builds. DO NOT CHANGE THE INCLUDE ORDER! */
+#define _CRTDBG_MAP_ALLOC
+#include <stdlib.h>
+#include <crtdbg.h>
+
+#include "c11threads.h"
+
+#include <assert.h>
+#include <limits.h>
+#include <malloc.h>
+#include <memory.h>
+#include <stddef.h>
+
+/* Condition variables and one-time callables need at least Windows Vista. */
+#ifdef C11THREADS_SUPPORT_WINNT_OLDER_THAN_VISTA
+#define WINVER 0x0400 /* Windows NT 4.0 */
+#define _WIN32_WINNT WINVER
+#endif
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+
+#if !defined(C11THREADS_SUPPORT_WINNT_OLDER_THAN_VISTA) && _WIN32_WINNT < 0x0600 /* Windows Vista */
+#error c11threads: Cannot support condition variables and call once on Windows older than Vista; define C11THREADS_SUPPORT_WINNT_OLDER_THAN_VISTA or C11THREADS_PTHREAD_WIN32 (use libpthread)
+#endif
+
+
+/* ---- globals ---- */
+
+struct _thrd_entry_win32_t {
+	thrd_t thrd;
+	HANDLE h;
+	struct _thrd_entry_win32_t *next;
+};
+
+struct _tss_dtor_entry_win32_t {
+	tss_t key;
+	tss_dtor_t dtor;
+	struct _tss_dtor_entry_win32_t *next;
+};
+
+_Bool _thrd_globals_initialized_win32 = 0;
+LARGE_INTEGER _thrd_perf_freq_win32;
+CRITICAL_SECTION _thrd_list_critical_section_win32;
+struct _thrd_entry_win32_t *_thrd_list_win32 = NULL;
+CRITICAL_SECTION _tss_dtor_list_critical_section_win32;
+struct _tss_dtor_entry_win32_t *_tss_dtor_list_win32 = NULL;
+
+static int __stdcall _thrd_init_globals_callback_win32(INIT_ONCE *init_once, void *parameter, void **context)
+{
+	(void)init_once;
+	(void)parameter;
+	(void)context;
+	QueryPerformanceFrequency(&_thrd_perf_freq_win32);
+#pragma warning(suppress: 28125)
```

Comment:
Put MSVC pragmas behind an ifdef. But in this case I'd just say drop them altogether. Users can tweak their warning levels to avoid annoying warnings.

Reply:
I'll be guarding them from non-Microsoft compilers.